parser/src/string.rs

```diff
@@ -0,0 +1,182 @@
+                    _ => Err(LexicalError {
+                        location,
+                        error: LexicalErrorType::OtherError(
+                            "Unexpected expression kind in string concatenation.".to_string(),
+                        ),
+                    }),
+                })
+                .flat_map(|result| match result {
+                    Ok(values) => values.into_iter().map(Ok).collect(),
+                    Err(e) => vec![Err(e)],
+                })
+                .collect::<Result<Vec<Expr>, LexicalError>>()?;
+
+            // De-duplicate adjacent constants.
+            let mut deduped: Vec<Expr> = vec![];
+            let mut current: Vec<String> = vec![];
+            for value in values {
+                match value.node {
+                    ExprKind::FormattedValue { .. } => {
+                        if !current.is_empty() {
+                            deduped.push(Expr::new(
+                                location,
+                                ExprKind::Constant {
+                                    value: Constant::Str(current.join("")),
+                                    kind: kind.clone(),
+                                },
+                            ));
+                            current.clear();
+                        }
+                        deduped.push(value)
+                    }
+                    ExprKind::Constant { value, .. } => {
+                        if let Constant::Str(value) = value {
+                            current.push(value);
+                        } else {
+                            return Err(LexicalError {
+                                location,
+                                error: LexicalErrorType::OtherError(
+                                    "Unexpected non-string constant.".to_string(),
+                                ),
+                            });
+                        }
+                    }
+                    _ => {
+                        return Err(LexicalError {
+                            location,
+                            error: LexicalErrorType::OtherError(
+                                "Unexpected expression kind in string concatenation.".to_string(),
+                            ),
+                        });
+                    }
+                }
+            }
+
+            if !current.is_empty() {
+                deduped.push(Expr::new(
+                    location,
+                    ExprKind::Constant {
+                        value: Constant::Str(current.join("")),
+                        kind,
+                    },
+                ));
+                current.clear();
+            }
+
+            Expr::new(location, ExprKind::JoinedStr { values: deduped })
+        }
+    } else {
+        values.into_iter().next().unwrap()
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::parser::parse_program;
+
+    #[test]
+    fn test_parse_string_concat() {
```

Comment:
These arguably belong in `parser.rs` since they depend on the parser (which depends on this file, rather than the other way around).